### PR TITLE
Branch grade list

### LINKED ISSUE #45: Create GradeList in accordance to the new UML
Refer to #36

### PR DESCRIPTION
**Pull request by RCPilot1604**
Fix #45 
Add JUnit tests to classes

Add GradeList 
Clean up UML
Implement GradeNotFound and DuplicateGrade Exceptions

### CODE REVIEW

File: src/main/java/tutorlink/lists/GradeList.java [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+        for (Grade grade : gradeArrayList) {
+            if (grade.getMatricNumber().equals(matricNumber) && grade.getComponentDescription().equals(componentDescription)) {
+                return studentArrayList.remove(grade);
+            }
+        }
+        return false;
+    }
+
+    public void addGrade(Grade grade) throws DuplicateGradeException {
+        for (Student gradeToCompare : gradeArrayList) {
```
Comment (jinzihan2002) [inline src/main/java/tutorlink/lists/GradeList.java]: For gradeToCompare it should be of type Grade?

Comment (RCPilot1604) [reply src/main/java/tutorlink/lists/GradeList.java]: Good catch, thanks.

File: src/main/java/tutorlink/lists/GradeList.java [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+                    && grade.getComponent().getName().equals(componentDescription)) {
+                return gradeArrayList.remove(grade);
+            }
+        }
+        return false;
+    }
+
+    public void addGrade(Grade grade) throws DuplicateGradeException {
+        for (Grade gradeToCompare : gradeArrayList) {
+            if (grade == gradeToCompare) {
```
Comment (ThienDuc3112) [inline src/main/java/tutorlink/lists/GradeList.java]: Use .equals or grade.isSameSubmission (which one to use depend command implementation)

 == in java only compared by references when using on non primitive types

File: src/main/java/tutorlink/lists/GradeList.java [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+        filteredList.gradeArrayList = gradeArrayList
+                .stream()
+                .filter(grade -> grade.getStudent().getMatricNumber().equals(matricNumber.toUpperCase())
+                && grade.getComponent().getName().equals(componentDescription.toLowerCase()))
+                .collect(Collectors.toCollection(ArrayList::new));
+        if (filteredList.gradeArrayList.isEmpty()) {
+            throw new GradeNotFoundException(String.format(ERROR_NO_GRADE_FOUND, matricNumber, componentDescription));
+        }
+        return filteredList;
+    }
```
Comment (ThienDuc3112) [inline src/main/java/tutorlink/lists/GradeList.java:62]: We could make it return all the student grades of a component if matric is empty, or return all component grade a student has if the component is empty.

Depend on whether we want that feature in our command or not.

**Review by ThienDuc3112 — CHANGES_REQUESTED** [review]
LGTM, putting as request change only because of == might cause a bug after a bit.

File: src/main/java/tutorlink/lists/StudentList.java [hunk 4]
```diff
@@ -10,9 +9,10 @@
 import java.util.stream.Collectors;
 import java.util.stream.IntStream;
 
-public class StudentList extends ItemList {
+public class StudentList {
     private static final String ERROR_DUPLICATE_MATRIC_NUMBER_ON_ADD = "Error! Student with Matric Number %s already"
             + "exists in the list!";
+    private static final String toStringDelimiter = "\n\t";
```
Comment (jinzihan2002) [inline src/main/java/tutorlink/lists/StudentList.java]: Should the constant be TO_STRING_DELIMITER ?

**Review by jinzihan2002 — COMMENTED** [review]
Overall LGTM, just minor naming convention to change

File: src/main/java/tutorlink/lists/GradeList.java (hunk 3 above)
Comment (RCPilot1604) [reply src/main/java/tutorlink/lists/GradeList.java:62]: Ok will merge first and implement this later.

Comment (RCPilot1604) [reply src/main/java/tutorlink/lists/GradeList.java:62]: Alright, thanks.

File: src/main/java/tutorlink/lists/StudentList.java (hunk 4 above)
Comment (RCPilot1604) [reply src/main/java/tutorlink/lists/StudentList.java]: Good catch, thanks.

**Review by ThienDuc3112 — APPROVED** [review]
LGTM

**Pull request merged**